### PR TITLE
Refactor of Metadata parsing in Bindings

### PR DESCRIPTION
**Pull request by berndverst**
# Description

Refactor of metadata parsing in Bindings.

I also added the ability to specify that metadata only applies to certain component types.

Example:

```go
type ComponentMetadata struct {
  MyField `mapstructure:"somefield" only:"pubsub,binding"`
}
```

This would only include this field "somefield" in Pubsub and Binding components metadata, but not any other type.

### CODE REVIEW

**Comment by berndverst** [comment]
FYI this PR didn't touch the components that failed, so it should be good to merge if it passes code review!

File: bindings/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -66,22 +74,23 @@ func (k *kubernetesInput) Init(ctx context.Context, metadata bindings.Metadata)
...
-	if val, ok := metadata.Properties["resyncPeriodInSec"]; ok && val != "" {
-		intval, err := strconv.ParseInt(val, 10, 64)
-		if err != nil {
-			k.logger.Warnf("invalid resyncPeriodInSec %s; %v; defaulting to 10s", val, err)
-			k.resyncPeriod = time.Second * 10
+func (k *kubernetesInput) parseMetadata(meta bindings.Metadata) error {
+	m := kubernetesMetadata{}
+	err := metadata.DecodeMetadata(meta.Properties, &m)
+	if err != nil {
+		if strings.Contains(err.Error(), "decoding 'resyncPeriodInSec': time: invalid duration") {
```
Comment (ItalyPaleAle) [inline bindings/kubernetes/kubernetes.go:81]: I would change this to:

```suggestion
		if strings.Contains(err.Error(), "resyncPeriodInSec") {
```

This is so if the upstream library (mapdecode) changes how it formats errors, this is less likely to break

File: metadata/utils.go [hunk 2]
```diff
@@ -204,9 +204,35 @@ func toStringArrayHookFunc() mapstructure.DecodeHookFunc {
 	}
 }
 
+type ComponentTypeValue string
+
+var ComponentType = struct {
```
Comment (ItalyPaleAle) [inline metadata/utils.go:209]: Does this need to be a struct? Can it be a set of constants instead?

As per our conversation on a recent PR, using a struct increases the memory usage, while consts should be easier on the memory.

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: metadata/utils.go (hunk 2 above)
Comment (berndverst) [reply metadata/utils.go]: Changed it to constants with an addition `switch` statement check.

**Review by ItalyPaleAle — APPROVED** [review]

**Review by ItalyPaleAle — APPROVED** [review]

**Comment by ekjuanrejon** [comment]
hello, does the zeebee component support basic authentication?

**Pull request merged**